Adding EfficientNetsB0-B7 support

masadcv opened the pull request:
Signed-off-by: masadcv <muhammad.asad@kcl.ac.uk>

Fixes #1090  .

### Description
Adding support for original EfficientNetB0-B7 models from: https://arxiv.org/abs/1905.11946

Based on https://github.com/lukemelas/EfficientNet-PyTorch
- Extended to use MONAI factories to generalise for spatial_dims in [1, 2, 3]
- Can be used with pretrained weights from above repo when spatial_dims=2
- Also adds a memory-efficient version of Swish activation function from: https://github.com/lukemelas/EfficientNet-PyTorch/issues/18#issuecomment-511677853 which is reported to save 30% memory during training on NVIDIA GPUs
- Added unittests to cover checking output shapes for EfficientnetB0-B7 implementation variations and correct loading of pretrained weights from above repo
- Added unittests to cover saving/loading with Torchscript using tests from `tests.utils.test_script_save()`
- Added a kitty test image for classification models unittesting at: ./tests/testing_data/kitty_test.jpg (image from: https://commons.wikimedia.org/wiki/File:Tabby_cat_with_blue_eyes-3336579.jpg)
- Added unittests to cover loading and running from pretrained weights for correct 'kitty' classification
- Fixed formatting/coding style to comply with ./runtests.sh
- Updated docs by adding new modules and checking the compiled html

### Status
**Ready**

### Types of changes
<!--- Put an `x` in all the boxes that apply, and remove the not applicable items -->
- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

wyli commented on tests/test_efficientnet.py:
```diff
@@ -0,0 +1,229 @@
+            "in_channels": 3,
+            "num_classes": 1000,
+        },
+        os.path.join(os.path.dirname(__file__), "testing_data/kitty_test.jpg"),
```
the path needs to be revised for paltform compatibility. cool image, is it with a CC-by license...
```suggestion
        os.path.join(os.path.dirname(__file__), "testing_data", "kitty_test.jpg"),
```

masadcv replied:
Thank you, I will update the code.
The image comes from reddit: https://www.reddit.com/r/aww/comments/h689o/kitten_with_a_bow_tie_xpost_from_pics/
Not entirely sure about the license. If you want, I can find and update to a CC-by license picture for a cat. Please feel free to suggest if you have any suggestions.

wyli replied:
thanks, public domain image as testing data would be great

masadcv replied:
I have updated the picture with the following that has CC0 License:
https://commons.wikimedia.org/wiki/File:Tabby_cat_with_blue_eyes-3336579.jpg

wyli commented on tests/test_efficientnet.py:
```diff
@@ -0,0 +1,229 @@
+            "in_channels": 3,
+            "num_classes": 1000,
+        },
+        os.path.join(os.path.dirname(__file__), "testing_data/kitty_test.jpg"),
```
```suggestion
        os.path.join(os.path.dirname(__file__), "testing_data", "kitty_test.jpg"),
```

wyli commented on monai/networks/layers/factories.py:
```diff
@@ -256,6 +256,13 @@ def swish_factory():
     return Swish
 
 
+@Act.factory_function("memswish")
```
could you please add a test in https://github.com/Project-MONAI/MONAI/blob/master/tests/test_activations.py#L51

masadcv replied:
This will be addressed in next commit.

wyli commented on monai/networks/nets/efficientnet.py:
```diff
@@ -0,0 +1,805 @@
+        blocks_args = _decode_block_list(blocks_args_str)
+
+        # checks for successful decoding of blocks_args_str
+        assert isinstance(blocks_args, list), "blocks_args should be a list"
```
the project's style is to only use `assert` in the test files,
so, this could be (please revise the other few places)
```py
if not isinstance(blocks_args, list):
    raise ValueError...
```

masadcv replied:
Thanks, I will resolve this in the next commit.

wyli commented on monai/networks/nets/efficientnet.py:
```diff
@@ -0,0 +1,805 @@
+            for _ in range(block_args.num_repeat - 1):
+                drop_connect_rate = drop_connect_rate
+                if drop_connect_rate:
+                    drop_connect_rate *= float(idx) / num_blocks  # scale drop connect_rate
```
codecov says this `drop_connect_rate` is not tested?

masadcv replied:
Thanks! Coincidently I have spotted a logical error here, which would have affected training. This logical error is addressed in next commit. I will think about how to isolate and test `drop_connect_rate`. The `drop_connect` layer itself can certainly be added to unittests, I will add this in next commit.

wyli reviewed (commented):
thanks @masadcv! this looks nice I have some minor comments inline, perhaps @yiheng-wang-nv could help double-check with the network details? also please add `@skip_if_quick`  to the unit tests if they requires downloading, e.g. https://github.com/Project-MONAI/MONAI/blob/master/tests/test_ahnet.py#L138

masadcv commented:
> thanks @masadcv! this looks nice I have some minor comments inline, perhaps @yiheng-wang-nv could help double-check with the network details? also please add `@skip_if_quick` to the unit tests if they requires downloading, e.g. https://github.com/Project-MONAI/MONAI/blob/master/tests/test_ahnet.py#L138

@wyli thank you for your comments! I have addressed all but the `drop_connect_rate` test. There doesn't seem to be a straightforward way to test it without exposing unnecessary arguments for the class. I have added unittests for `drop_connect` layer which covers a number of cases. The unittests with pretrained models now use `@skip_if_quick` as well as test url loading using `test_pretrained_networks`.  We may also want to look into hosting models on a mirror in case we loose the ones from original repo. 

The kitty test image has been updated to use a Creative Commons image from wikimedia commons. 

I need a few more passes to have everything double-checked and in place. I will also run the full coverage integration tests as soon as I have access to relevant resources. 

@yiheng-wang-nv would it be possible to have your review in the coming days? Thank you!

yiheng-wang-nv commented on tests/test_efficientnet.py:
```diff
@@ -0,0 +1,285 @@
+    return ["efficientnet-b{}".format(d) for d in range(8)]
+
+
+def get_expected_model_shape(model_name):
```
Could you please also add test cases for different input shapes?

masadcv replied:
What do you mean by this? The input shapes vary with each version of the network. All tests have been implemented with respective input shapes for the networks. See https://github.com/Project-MONAI/MONAI/blob/d83a053f1c14ea4f90787bef1be4263e00076d5c/tests/test_efficientnet.py#L79 where this is done.

yiheng-wang-nv replied:
I mean for actual situation, the network should also work for other shapes, rather than only the original shapes, like [192, 192].

masadcv replied:
Okay, I will setup some tests for it. Do you have any suggestions of nice shapes to cover for medical imaging related problems?

masadcv replied:
> I mean for actual situation, the network should also work for other shapes, rather than only the original shapes, like [192, 192].

This has been done now for shapes in [128, 256, 512] - all of which are non-default to the networks:

https://github.com/Project-MONAI/MONAI/blob/45c2d91b1554354dbc1ccbd1016ba407554a2d64/tests/test_efficientnet.py#L204

yiheng-wang-nv commented:
Thanks @masadcv , I added some comments above.
Hi @wyli , does efficientnet need to support torch script?

wyli commented:
could you please add test_efficientnet to the "min_tests" list of excluding cases https://github.com/Project-MONAI/MONAI/blob/d56c002c70bfccc60a655cc160bd7293e54e7eff/tests/min_tests.py#L29 and fix/remove some of the problematic type hints, I'll make sure this PR is included in v0.5, many thanks @masadcv !

masadcv commented on monai/networks/nets/efficientnet.py:
```diff
@@ -0,0 +1,827 @@
+        return nn.Identity()
+
+
+def _decode_block_list(string_list):
```
done

masadcv commented on monai/networks/nets/efficientnet.py:
```diff
@@ -0,0 +1,827 @@
+    return _paddings_ret
+
+
+def _make_same_padder(conv_op, image_size: List[int]):
```
I have added `Type[Union[nn.Conv1d, nn.Conv2d, nn.Conv3d]]` which should suffice for efficientnet.
In future, if we want to use this for doing Tensorflow style `SAME` padding for other operations (e.g. MaxPoolNd etc) then we can extend to use in those cases.

masadcv commented on monai/networks/nets/efficientnet.py:
```diff
@@ -0,0 +1,827 @@
+    return output
+
+
+def _calculate_output_image_size(input_image_size, stride: Union[int, Tuple[int]]):
```
done

masadcv commented on monai/networks/nets/efficientnet.py:
```diff
@@ -0,0 +1,827 @@
+    return int(new_filters)
+
+
+def _round_repeats(repeats, depth_coefficient=None):
```
done

masadcv commented on monai/networks/nets/efficientnet.py:
```diff
@@ -0,0 +1,827 @@
+    return res
+
+
+def _round_filters(filters, width_coefficient=None, depth_divisor=None):
```
done

masadcv commented on monai/networks/nets/efficientnet.py:
```diff
@@ -0,0 +1,827 @@
+        stride: Union[int, Tuple[int]],
+        image_size: List[int],
+        expand_ratio: int,
+        se_ratio: float,
```
done, also added `Optional` to other places where it was needed

masadcv commented on monai/networks/nets/efficientnet.py:
```diff
@@ -0,0 +1,827 @@
+    Returns:
+        output: output tensor after applying drop connection.
+    """
+    if p < 0 or p > 1:
```
done

masadcv commented:
> could you please add test_efficientnet to the "min_tests" list of excluding cases
> 
> https://github.com/Project-MONAI/MONAI/blob/d56c002c70bfccc60a655cc160bd7293e54e7eff/tests/min_tests.py#L29
> and fix/remove some of the problematic type hints, I'll make sure this PR is included in v0.5, many thanks @masadcv !

@wyli many thanks! I have done both, locally all tests are passing. It would be awesome to have this in v0.5. Once that is done, I will also add a tutorial. Do you have any suggestions for medical imaging related classification dataset that I can use for a simple tutorial?

@yiheng-wang-nv many thanks for your comments, I have addressed most of your comments. I will look into testing different shapes as you suggested.

masadcv commented on monai/networks/nets/efficientnet.py:
```diff
@@ -0,0 +1,827 @@
+        batch_norm_momentum: float = 0.99,
+        batch_norm_epsilon: float = 1e-3,
+        drop_connect_rate: float = 0.2,
+        depth_divisor=8,
```
done

masadcv commented on monai/networks/nets/efficientnet.py:
```diff
@@ -0,0 +1,827 @@
+
+        Args:
+            inputs (tensor): Input tensor.
+            drop_connect_rate (bool): Drop connect rate (float, between 0 and 1).
```
thanks for pointing this out. I have removed it.

wyli commented:
there's still an error here: https://github.com/Project-MONAI/MONAI/pull/1938/checks?check_run_id=2296192348#step:6:7760

tutorial -- we have a demo script using the IXI dataset (https://brain-development.org/ixi-dataset/) for gender classification https://github.com/Project-MONAI/tutorials/blob/master/3d_classification/torch/densenet_training_array.py perhaps you can make a new example based on it

masadcv commented:
> there's still an error here: https://github.com/Project-MONAI/MONAI/pull/1938/checks?check_run_id=2296192348#step:6:7760
> 
> tutorial -- we have a demo script using the IXI dataset (https://brain-development.org/ixi-dataset/) for gender classification https://github.com/Project-MONAI/tutorials/blob/master/3d_classification/torch/densenet_training_array.py perhaps you can make a new example based on it

I am looking into this error now. Should resolve soon.
Thanks for the pointing me in the direction of tutorials.

wyli commented on tests/test_efficientnet.py:
```diff
@@ -0,0 +1,285 @@
+            self.assertEqual(result_shape, expected_shape)
+
+    def test_script(self):
+        net = EfficientNetBN(model_name="efficientnet-b0", spatial_dims=2, in_channels=3, num_classes=1000)
```
@yiheng-wang-nv this one mostly supports torchscript, are we good here?

yiheng-wang-nv replied:
Yes, I think this case is good!

wyli reviewed (approved):
thanks, it looks nice!

yiheng-wang-nv commented:
> > could you please add test_efficientnet to the "min_tests" list of excluding cases
> > https://github.com/Project-MONAI/MONAI/blob/d56c002c70bfccc60a655cc160bd7293e54e7eff/tests/min_tests.py#L29
> > 
> > and fix/remove some of the problematic type hints, I'll make sure this PR is included in v0.5, many thanks @masadcv !
> 
> @wyli many thanks! I have done both, locally all tests are passing. It would be awesome to have this in v0.5. Once that is done, I will also add a tutorial. Do you have any suggestions for medical imaging related classification dataset that I can use for a simple tutorial?
> 
> @yiheng-wang-nv many thanks for your comments, I have addressed most of your comments. I will look into testing different shapes as you suggested.

Many thanks @masadcv , this PR looks nice!

masadcv commented:
Seems to be failing on unrelated test now:
https://github.com/Project-MONAI/MONAI/pull/1938/checks?check_run_id=2297721328
